File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.java
class Solution {
    public int firstStableIndex(int[] nums, int k) {
        int n = nums.length;

        int[] M = new int[n];
        M[0] = nums[0];
        for (int i=1; i<n; i++) {
            M[i] = Math.max (M[i-1], nums[i]);
        }

        int[] m = new int[n];
        m[n-1] = nums[n-1];
        for (int i=n-2; i>=0; i--) {
            m[i] = Math.min (m[i+1], nums[i]);
        }

        for (int i=0; i<n; i++) {
            if (M[i] - m[i] <= k) return i;
        }

        return -1;
    }
}

File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        n = len(nums)

        M = [0] * n
        M[0] = nums[0]
        for i in range (1,n):
            M[i] = max(nums[i], M[i-1])

        m = [0] * n
        m[n-1] = nums[n-1]
        for i in range (n-2, -1, -1):
            m[i] = min (m[i+1], nums[i])

        for i in range (n):
            if M[i] - m[i] <= k:
                return i

        return -1